src/threeIdProvider.ts

```diff
@@ -8,105 +8,111 @@ import {
   createHandler,
 } from 'rpc-utils'
 
-import { EncryptedMessage } from './crypto'
-import IdentityWallet from './identity-wallet'
+import { sha256Multihash, pad, unpad } from './utils'
+import { didMethods, ProviderConfig, Context } from './did-provider'
+import { PublicKeys } from './keyring'
 
 type Origin = string | null | undefined
 
-type Context = {
-  provider: ThreeIdProvider
-  origin: Origin
-}
-
 const methods: HandlerMethods<Context> = {
-  '3id_getLink': async (ctx) => {
-    return (await ctx.provider.wallet.getLink()).toLowerCase()
-  },
-  '3id_linkManagementKey': async (ctx) => {
-    return await ctx.provider.wallet.linkManagementKey()
-  },
-  '3id_authenticate': async (ctx, params) => {
-    return await ctx.provider.wallet.authenticate(
-      params.spaces,
-      {
-        authData: params.authData,
-        address: params.address,
-        mgmtPub: params.mgmtPub,
-      },
-      ctx.origin
-    )
+  '3id_authenticate': async ({ permissions, keyring, origin }, params) => {
+    const spaces = await permissions.request(origin, params.spaces || [])
+    if (spaces === null) throw new RPCError(4001, 'User Rejected Request')
+    return {
+      main: keyring.getPublicKeys({ mgmtPub: params.mgmtPub }),
+      spaces: spaces.reduce((acc, space) => {
+        acc[space] = keyring.getPublicKeys({
+          space,
+          uncompressed: true,
+        })
+        return acc
+      }, {} as Record<string, PublicKeys>),
+    }
   },
-  '3id_isAuthenticated': async (ctx, params) => {
-    return await ctx.provider.wallet.isAuthenticated(params.spaces, ctx.origin)
+  '3id_isAuthenticated': async ({ permissions, origin }, params) => {
+    return Promise.resolve(permissions.has(origin, params.spaces))
```

Comment:
Not an issue but it seems this function doesn't have to be `async` / using `Promise.resolve`

Reply:
Doesn't it have to since the `send` method awaits the `_handle` method?

Reply:
AFAIK the way `await` is implemented it checks if the result is a promise or not to extract the value, so it works either way, but might be safer not to rely on this behavior anyways.